### DISCUSSION TITLE
Document API with OpenApi 3.0 specs

### DISCUSSION THREAD

**Discussion by nasoblu5** [PostgREST/postgrest#2673] [Q&A] [answered]
I've seen this topic has been touched some year ago but documentation of the API in POSTGREST using OpenApi 3.0 appears still unsolved.

Is there some hope in a next future to be applied in POSTGREST ? because people complain about 

Anyway I see OpenApi.hs component doesn't solve neither GET and POST for the same stored procedure with overloaded parameters as it should be in OpenApi 2.0 specs.

Indeed I have the procedured Requests declared one time as IMMUTABLE with an integer argument and another as VOLATILE with JSON argument. 

On landing page of POSTGREST http://localhost/3000 only the POST API has been document, no track for GET API of the same procedure.

Is there a workaround for this issue ?

Thanks

**Comment by laurenceisla** [comment, marked as answer]
> Is there some hope in a next future to be applied in POSTGREST ? because people complain about

Hi. Yes, there's this issue https://github.com/PostgREST/postgrest/issues/1698 that explains the decision. In summary, we're planning to remove the OpenAPI implementation from core and move it to a separate repository in a language that promotes more collaboration from the community. In the meantime we're resolving some minor issues that may present here and there, but not a major overhaul.

> Anyway I see OpenApi.hs component doesn't solve neither GET and POST for the same stored procedure with overloaded parameters as it should be in OpenApi 2.0 specs.

I think this is not allowed by the specification as mentioned here https://github.com/OAI/OpenAPI-Specification/issues/164, or maybe I'm misunderstanding something. I don't know if there is a workaround, it looks like that only a different name for each function would solve this for now.

> On landing page of POSTGREST http://localhost/3000 only the POST API has been document, no track for GET API of the same procedure.

This was fixed in #2651. The version [10.2.0.20230209](https://github.com/PostgREST/postgrest/releases/tag/v10.2.0.20230209)  has that change implemented.

**Reply by nasoblu5** [reply]
Maybe a workaround could be replacing Postgrest landing page http://localhost:3000/ with alternative JSON generated content using Swagger editor and mapping the root endpoint to a user's stored procedure ?

**Reply by steve-chavez** [reply]
@nasoblu5 Actually that already exists but it's undocumented because it's experimental. Try setting the `db-root-spec` config to a function.

**Reply by nasoblu5** [reply]
Thanks Steve,

it works fine, I used the db configuration prgst.db_root_spec="swagger" where swagger is a path to stored procedure that returns a JSON compiled using Swagger Editor 3.0 compliant with OpenApi 3.0 specs.

It should be perfect if it is possible to return a yaml document instead JSON.

**Reply by steve-chavez** [reply]
Cool!

> where swagger is a path to stored procedure that returns a JSON compiled using Swagger Editor 3.0 compliant with OpenApi 3.0 specs

Do you have an example of this you could share?

> It should be perfect if it is possible to return a yaml document instead JSON.

We're working on accepting custom media types on https://github.com/PostgREST/postgrest/pull/1582.

**Reply by nasoblu5** [reply]
Steve,

It is very trivial, I have a table "configurations" for various settings and I associated to a key the JSON MANUALLY compiled with Swagger editor with work in progress. Sadly with Postgrest I can manage overloaded parameters and we can have several GETs for the same route in the schema cache, instead this is forbidden in OpenApi 3.0 and you should rename additional GETs to the first and only. 

Anyway this is the route:

CREATE OR REPLACE FUNCTION rest.swagger()
    RETURNS json
    LANGUAGE 'plpgsql'
    COST 100
    IMMUTABLE SECURITY DEFINER PARALLEL UNSAFE
AS $BODY$
DECLARE
jValue json;
BEGIN
  SELECT value::json
  INTO   jValue  
  FROM   portal.configurations
  WHERE  key = 'SWAGGER_TEMPLATE'; 
  
  RETURN jValue;   
END
$BODY$;
